File: PHP/test/HeroQuestTest.php
<?php

namespace Tests;

use PHPUnit\Framework\TestCase;
use CodingDojo\HeroQuest;

class HeroQuestTest extends TestCase {

    private $playerName;
    private $playerHealth;
    private $playerStrength;
    private $playerMagic;
    private $playerCraftingSkill;
    private $itemName;
    private $itemKind;
    private $itemPower;

    protected function setUp(): void {
        $this->playerName = "Conan";
        $this->playerHealth = 100;
        $this->playerStrength = 20;
        $this->playerMagic = 10;
        $this->playerCraftingSkill = 10;
        $this->itemName = "Amulet of Strength";
        $this->itemKind = "Strength";
        $this->itemPower = 10;
    }

    public function testPlayerToString() {
        $result = HeroQuest::playerToString(
            $this->playerName,
            $this->playerHealth,
            $this->playerStrength,
            $this->playerMagic,
            $this->playerCraftingSkill
        );

        $expected = "Conan's Attributes:\nHealth: 100\nStrength: 20\nMagic: 10\nCrafting Skill: 10\n";
        $this->assertEquals($expected, $result);
    }

    public function testPlayerFallsDown() {
        $this->playerStrength = 3;
        HeroQuest::playerFallsDown($this->playerHealth, $this->playerStrength);
        $this->assertEquals(90, $this->playerHealth);
    }

    public function testPlayerFallsDownNoDamage() {
        HeroQuest::playerFallsDown($this->playerHealth, $this->playerStrength);
        $this->assertEquals(100, $this->playerHealth);
    }

    public function testItemToString() {
        $result = HeroQuest::itemToString(
            $this->itemName,
            $this->itemKind,
            $this->itemPower
        );

        $expected = "Item: Amulet of Strength\nKind: Strength\nPower: 10\n";
        $this->assertEquals($expected, $result);
    }

    public function testItemReduceByUsage() {
        HeroQuest::itemReduceByUsage($this->itemKind, $this->itemPower);
        $this->assertEquals(5, $this->itemPower);
    }

    public function testItemReduceByUsageToJunk() {
        $this->itemPower = 1;
        HeroQuest::itemReduceByUsage($this->itemKind, $this->itemPower);
        $this->assertEquals(0, $this->itemPower);
        $this->assertEquals("Junk", $this->itemKind);
    }

    public function testItemApplyEffectToPlayer() {
        HeroQuest::itemApplyEffectToPlayer(
            $this->itemName,
            $this->itemKind,
            $this->itemPower,
            $this->playerHealth,
            $this->playerStrength,
            $this->playerMagic
        );

        $this->assertEquals(30, $this->playerStrength);
    }

    public function testItemApplyEffectToPlayerJunk() {
        $this->itemKind = "Junk";
        HeroQuest::itemApplyEffectToPlayer(
            $this->itemName,
            $this->itemKind,
            $this->itemPower,
            $this->playerHealth,
            $this->playerStrength,
            $this->playerMagic
        );
        $this->assertEquals(20, $this->playerStrength);
    }

    public function testItemRepair() {
        HeroQuest::itemRepair($this->playerCraftingSkill, $this->itemPower);
        $this->assertEquals(26, $this->itemPower);
    }
}

File: go_struct/hero_quest/hero_quest_test.go
package hero_quest_test

import (
	"testing"

	"github.com/codecop/Hero-Quest-Refactoring-Kata-Struct/hero_quest"
	"github.com/stretchr/testify/assert"
)

func TestPlayerToString(t *testing.T) {
	questData := createQuestData()

	result := hero_quest.PlayerToString(questData.PlayerName, questData.PlayerHealth, questData.PlayerStrength, questData.PlayerMagic, questData.PlayerCraftingSkill)

	assert.Equal(t, "Conan's Attributes:\nHealth: 100\nStrength: 20\nMagic: 10\nCrafting Skill: 10\n", result)
}

func TestPlayerFallsDown(t *testing.T) {
	questData := createQuestData()
	questData.PlayerStrength = 3

	hero_quest.PlayerFallsDown(&questData)

	assert.Equal(t, 90, questData.PlayerHealth)
}

func TestPlayerFallsDownNoDamage(t *testing.T) {
	questData := createQuestData()

	hero_quest.PlayerFallsDown(&questData)

	assert.Equal(t, 100, questData.PlayerHealth)
}

func TestItemToString(t *testing.T) {
	questData := createQuestData()

	result := hero_quest.ItemToString(questData.ItemName, questData.ItemKind, questData.ItemPower)

	assert.Equal(t, "Item: Amulet of Strength\nKind: Strength\nPower: 10\n", result)
}

func TestItemReduceByUsage(t *testing.T) {
	questData := createQuestData()

	hero_quest.ItemReduceByUsage(&questData)

	assert.Equal(t, 5, questData.ItemPower)
}

func TestItemReduceByUsageToJunk(t *testing.T) {
	questData := createQuestData()
	questData.ItemPower = 1

	hero_quest.ItemReduceByUsage(&questData)

	assert := assert.New(t)
	assert.Equal(0, questData.ItemPower)
	assert.Equal("Junk", questData.ItemKind)
}

func TestItemApplyEffectToPlayer(t *testing.T) {
	questData := createQuestData()

	hero_quest.ItemApplyEffectToPlayer(&questData)

	assert.Equal(t, 30, questData.PlayerStrength)
}

func TestItemApplyEffectToPlayerJunk(t *testing.T) {
	questData := createQuestData()
	questData.ItemKind = "Junk"

	hero_quest.ItemApplyEffectToPlayer(&questData)

	assert.Equal(t, 20, questData.PlayerStrength)
}

func TestItemRepair(t *testing.T) {
	questData := createQuestData()

	hero_quest.ItemRepair(&questData)

	assert.Equal(t, 26, questData.ItemPower)
}

func createQuestData() hero_quest.QuestData {
	return hero_quest.QuestData{
		PlayerName:          "Conan",
		PlayerHealth:        100,
		PlayerStrength:      20,
		PlayerMagic:         10,
		PlayerCraftingSkill: 10,
		ItemName:            "Amulet of Strength",
		ItemKind:            "Strength",
		ItemPower:           10,
	}
}
